_posts/kevin_hartmann/2017-07-20-Why-Siri-Cant-Code.md

```diff
@@ -0,0 +1,484 @@
+
+ Vast libraries of existing 
+code let us re-use known solutions to an assortment of problems; and IDEs and syntax checkers help give us hints and assistance; and various syntax checkers and
+proof verifiers already exist, and are getting better.
+
+ Isn't that enough?
+
+### It's not enough; it's never enough!
+<image src="{{ site.urlimg }}/kevin_hartmann/never_enough4.jpg" /> 
+
+For all that they can do, our existing tools take a very limited
+approach to understanding, verifying, or validating our code. Most of them focus on solving specific problems that are known to occur within a given programming language, such as Findbugs for Java, or Valgrind for C. Our tools work, but for the most part, they can't be said to be very "intelligent"; at least, not in the 
+sense of understanding what the code they generate is supposed to do, or how it's supposed to work. 
+
+Most of the hard part, making sure that the code that our tools produce not only works, but works correctly, and works according to our requirements is left to us, the programmers; and coding errors are so common that we 
+typically rely on entire "QA" departments to help us find and fix these flaws; making mistakes is something that just happens. Our tools catch a few of our mistakes; but they miss a lot more of them. That's not an easy problem to fix.
+
+## Arguing about semantics
+<image src="{{ site.urlimg }}/kevin_hartmann/semantic image 1.gif" /> 
+
+
+One thing that our current tools are lacking is an understanding of what computer scientists call **_semantics_**; the underlying _meaning_ of what our programs do. Instead, most of our tools focus on **_syntax_**; the _representation_ of our code as patterns of symbols. When a programmer defines a programming language, they typically define the syntax very carefully and explicitly, using a mark-up language for syntax such as [BNF](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form); but the language's semantics are usually written as an ad-hoc series of rules written in a human language, such as English, if they are even written down at all.
+
+We leave it up to the writer of a compiler or interpreter to correctly translate the meaning of a program from it's syntactic inputs into code whose semantic meaning matches that of the programming language we're writing in. This job isn't an easy or an obvious one.
+
+
+If we had tools that were capable of understanding and manipulating the code at these deeper levels of semantic understanding, we could automatically verify properties of programs, and even automatically generate correct compilers don't even exist yet! 
```

Comment:
Thanks! The problem with reading your own writing is that you already know what it was supposed to say, so that's what you see when you read it. ;-)